### ISSUE TITLE
Visual Studio Code Freezes and causes High CPU Usage

### ISSUE THREAD

**Issue by MainaWycliffe** [HvyIndustries/crane#300]
Platform : win32
VSCode Version : 1.11.2
Crane Version : 0.3.6
---
When i use your extension, it causes Visual Studio Code to freeze, stops intelliSense from working, sometime i even have to restart the application. Also during that period, the cpu usage is high. i have narrowed it down to your extension since when i disable it, the problem goes away.

**Comment by ichiriac** [comment]
Hi @MainaWycliffe,

Can you tell us how many PHP files do you have, and do you have some files with big size, more than 500Ko. Also could help to know if you use a framework that is located in the same folder as your scripts. Another thing that could make the parser slow could be that fact you are working with files located on a network drive.

In principle crane scans the project once, and then only parses the current working file. The only edge case that could block everything could be an infinite loop on a file that contains a syntax error. 

If it's the case what could help us is to have the source code of this file, you can locate it by enabling the debug output and send us the contents of the file that locks the process.

**Comment by MainaWycliffe** [comment]
Hi @ichiriac 

I have about 1000 php pages, most of them are small files of between 1 to 5kbs although the project contains other file types such as HTML, CSS and Javascript. The total size of the php files alone comes at about 5Mbs. I am not using any php framework.

**Comment by ichiriac** [comment]
Thanks  @MainaWycliffe for the sample. I've checked it it doesn't contains any error, and even when I try to edit it from VSCode it does not triggers extra CPU parsing issues (or autocomplete). 

Could also provide classes that extends this one ? It's maybe not a parsing problem but maybe an infinite recusive class tree inspection problem for building autocompletion.

**Comment by ichiriac** [comment]
Ok, got it : 

```php
<?php
class Foo extends Bar {
  public function foo() {

  }
}
class Bar extends Foo {
  public function bar() {
    $this->
  }
}
```

Here the error message :

```
[ERROR] Completion error: Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at Function.getNamespaceInfoFromFQNClassname (/home/devbox/.vscode/extensions/HvyIndustries.crane-0.3.6/server/util/Namespaces.js:12:13)
    at SuggestionBuilder.getClassNodeFromTree (/home/devbox/.vscode/extensions/HvyIndustries.crane-0.3.6/server/suggestionBuilder.js:549:53)
    at SuggestionBuilder.addClassMembers (/home/devbox/.vscode/extensions/HvyIndustries.crane-0.3.6/server/suggestionBuilder.js:779:42)
    at SuggestionBuilder.addClassMembers (/home/devbox/.vscode/extensions/HvyIndustries.crane-0.3.6/server/suggestionBuilder.js:781:49)
    at SuggestionBuilder.addClassMembers (/home/devbox/.vscode/extensions/HvyIndustries.crane-0.3.6/server/suggestionBuilder.js:781:49)
```

**Comment by MainaWycliffe** [comment]
Could having a PHP SDK from AWS course the problem, because the extended file does not have much and i have used it extensively through the project

[model.php.txt](https://github.com/HvyIndustries/crane/files/931011/model.php.txt)
[aws_library.zip](https://github.com/HvyIndustries/crane/files/931004/aws_library.zip)

**Comment by ichiriac** [comment]
Hi  @nevadascout,

I've just published a minor issue release that contains a fix for this, in `2.0.4` the parser can handle static closures

It should be safe have run it on every framework without any regression and there is no backward change (just added new properties)

**Comment by nevadascout** [comment]
I've updated the php-parser version, but I'm now unable to reproduce any issues with the infinite loop using the `foo extends bar` and `bar extends foo` example code you provided
